lib/cdo/graphics/certificate_image.rb

```diff
@@ -15,31 +15,69 @@ def create_certificate_image2(image_path, name, params={})
   pointsize = 68
   x_offset = params[:x] || 0
   y_offset = params[:y] || 0
-  apply_text(background, name, pointsize, font, color, x_offset, y_offset)
+  # The area in pixels under the "Certificate of Completion" on the certificate reserved for the name.
+  name_area_width = 800
+  name_area_height = 80
+  apply_text(background, name, pointsize, font, color, x_offset, y_offset, name_area_width, name_area_height)
   background
 end
 
-# applies the given text to the given image object.
-def apply_text(image, text, pointsize, font, color, x_offset, y_offset)
+# Applies the given text to the given image object.
+# The text is shrunk to fit the given width and height.
+#
+# @param [Magick::Image] image is the background to put text on.
+# @param [String] text is the string to add to the image.
+# @param [Integer] pointsize is the max size to display the text at.
+# @param [String] font to use e.g. 'Helvetica bold'
+# @param [String] color of the text e.g. 'rgb(118,101,160)'
+# @param [Integer] x_offset of the position to center the text at.
+# @param [Integer] y_offset of the position to center the text at.
+# @param [Integer] width of the bounding box for the text.
+# @param [Integer] height of the bounding box for the text.
+def apply_text(image, text, pointsize, font, color, x_offset, y_offset, width=nil, height=nil)
   # If there is no text, don't try to render it.
   return if text.nil? || text.strip.empty?
+  # If there is no background image, there is nothing to do.
+  return if image.nil?
+
   text = escape_image_magick_string(text)
+  # If no bounding box given, default to the background image width & height.
+  width ||= image.columns
+  height ||= image.rows
 
   begin
     # The text will be put into an image with a transparent background.  This
     # uses 'pango', the OS's text layout engine, in order to dynamically select
     # the correct font. This is important for handling non-latin languages.
+    # The text_overlay is first generated at full size and then resized to fit
+    # the given bounding box width & height.
     text_overlay = Magick::Image.read("pango:#{text}") do
       # pango:markup is set to false in order to easily prevent pango markup
       # injection from user input.
       define('pango', 'markup', false)
+      # If the text doesn't fit the bounding box, then put a '...' at the end.
+      define('pango', 'ellipsize', 'end')
       self.background_color = 'none'
       self.pointsize = pointsize
       self.font = font
       self.fill = color
-      # Limit the size of the text_overlay to the size of the background image.
-      self.size = "#{image.columns}x#{image.rows}"
+      # Limit the maximum size of content to protect ourselves against extremely large strings.
+      self.size = "#{width * 3}x#{height * 3}"
```

Comment:
Are `width` and `height` pixels or characters?  Also, why `* 3`?

Reply:
Added a comment, let me know if it works for you.